src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -45,11 +45,14 @@ public JsonConverter GetConverter(Type typeToConvert)
                 ThrowHelper.ThrowArgumentNullException(nameof(typeToConvert));
             }
 
-            if (JsonSerializer.IsReflectionEnabledByDefault && _typeInfoResolver is null)
+            if (JsonSerializer.IsReflectionEnabledByDefault)
             {
                 // Backward compatibility -- root & query the default reflection converters
                 // but do not populate the TypeInfoResolver setting.
-                return DefaultJsonTypeInfoResolver.GetConverterForType(typeToConvert, this);
+                if (_typeInfoResolver is null)
```

Comment:
was the lack of this check a bug previously?

Reply:
See https://github.com/dotnet/runtime/pull/90013#discussion_r1284314974. Seems the previous structure was resulting in rooted reflection components in trimmed apps even though it's equivalent.